stores/filestore_test.go

```diff
@@ -2001,50 +2017,60 @@ func TestFSAutoSync(t *testing.T) {
 	storeMsg(t, c, "foo", 1, []byte("hello"))
 	storeSub(t, c, "foo")
 
-	time.Sleep(150 * time.Millisecond)
+	time.Sleep(50 * time.Millisecond)
 
-	ms.Lock()
-	notFlushed = ms.bw != nil && ms.bw.buf != nil && ms.bw.buf.Buffered() > 0
-	ms.Unlock()
-	if notFlushed {
-		t.Fatalf("Message store should have been flushed")
+	msSynced := checkMsgStoreFlushed(t, ms, true)
+	ssSynced := checkSubStoreFlushed(t, ss, true)
+
+	// Check that without new activity, there is no unnecessary sync'ing.
+	time.Sleep(100 * time.Millisecond)
+
+	if n := checkMsgStoreFlushed(t, ms, true); n != msSynced {
+		t.Fatalf("Message store is unnecessarily sync'ed (sync count was %v, now %v)", msSynced, n)
+	}
+	if n := checkSubStoreFlushed(t, ss, true); n != ssSynced {
+		t.Fatalf("Subscription store is unnecessarily sync'ed (sync count was %v, now %v)", ssSynced, n)
 	}
 
-	ss.Lock()
-	notFlushed = ss.bw != nil && ss.bw.buf != nil && ss.bw.buf.Buffered() > 0
-	ss.Unlock()
-	if notFlushed {
-		t.Fatalf("Message store should have been flushed")
+	// Add new activity and check things are now updated.
+	storeMsg(t, c, "foo", 2, []byte("hello"))
+	storeSub(t, c, "foo")
+
+	time.Sleep(50 * time.Millisecond)
+
+	if n := checkMsgStoreFlushed(t, ms, true); n == msSynced {
+		t.Fatalf("Message store was not sync'ed after new activity (sync count was %v, now %v)", msSynced, n)
+	}
+	if n := checkSubStoreFlushed(t, ss, true); n == ssSynced {
+		t.Fatalf("Subscription store was not sync'ed after new activity (sync count was %v, now %v)", ssSynced, n)
 	}
 
 	s.Close()
 
 	cleanupFSDatastore(t)
 	// Verify that auto sync works if a channel is removed
-	// Make the auto sync code wait for 300ms when getting ready to sync a msg/sub store.
-	atomic.StoreInt64(&testAutoSync, int64(300*time.Millisecond))
-	defer atomic.StoreInt64(&testAutoSync, 0)
-
 	s = createDefaultFileStore(t, BufferSize(1024), AutoSync(150*time.Millisecond))
 	defer s.Close()
 
-	c = storeCreateChannel(t, s, "foo")
+	channels := map[string]struct{}{}
+	for i := 0; i < 100; i++ {
+		cn := fmt.Sprintf("foo_%d", i)
+		c = storeCreateChannel(t, s, cn)
+		channels[cn] = struct{}{}
 
-	storeMsg(t, c, "foo", 1, []byte("hello"))
-	subID := storeSub(t, c, "foo")
-	storeSubDelete(t, c, "foo", subID)
-
-	// Wait for the autosync callback to pick up the channel's stores to
-	// perform an autosync. We have made the code wait 300ms before
-	// issuing the autoSync() call...
-	time.Sleep(200 * time.Millisecond)
+		storeMsg(t, c, cn, 1, []byte("hello"))
+		subID := storeSub(t, c, cn)
+		storeSubDelete(t, c, cn, subID)
+	}
 
-	// Now delete channel
-	if err := s.DeleteChannel("foo"); err != nil {
-		t.Fatalf("Error deleting channel: %v", err)
+	// Delete channels around the time the auto-sync would hit.
+	time.Sleep(140 * time.Millisecond)
```

Comment:
Feels too random for testing, but maybe its ok.

Reply:
To be honest, this test about DeleteChannel is not really relevant now that I don't do it at the FileStore level. Original code would grab all channels under FileStore lock, then release the lock and then individually auto-sync on each msg/sub store for each channel. The "risk" was that the channel had been deleted between the gather and the actual auto-sync. This can no longer happen since the auto-sync is brought down to the MsgStore and SubStore.

Reply:
Maybe delete the test then?

Reply:
that part of it yes.